solidity/random-beacon/test/RandomBeacon.Relay.test.ts

```diff
@@ -257,78 +403,332 @@ describe("RandomBeacon - Relay", () => {
+            // perform any signing so their public key can be arbitrary bytes.
+            // Also, that group is created just after the relay request is
+            // made to ensure it is not selected for signing the original request.
+            await (randomBeacon as RandomBeaconStub).roughlyAddGroup(
+              "0x01",
+              signersIDs
+            )
+
+            // `groupSize * relayEntrySubmissionEligibilityDelay +
+            // relayEntryHardTimeout`.
+            await mineBlocks(64 * 10 + 5760)
+
+            tx = await randomBeacon.reportRelayEntryTimeout()
+          })
+
+          it("should slash entire stakes of all group members", async () => {
+            await expect(tx)
+              .to.emit(staking, "Slashed")
+              .withArgs(to1e18(1000), signersAddresses)
+          })
+
+          it("should terminate the group", async () => {
+            // TODO: Implementation once `Groups` library is ready.
+          })
+
+          it("should emit RelayEntryTimedOut event", async () => {
+            await expect(tx)
+              .to.emit(randomBeacon, "RelayEntryTimedOut")
+              .withArgs(1)
+          })
+
+          it("should retry current relay request", async () => {
+            // We expect the same request ID because this is a retry.
+            // Group ID is still `0` because there is only one group
+            // after termination was performed.
+            await expect(tx)
+              .to.emit(randomBeacon, "RelayEntryRequested")
+              .withArgs(1, 0, blsData.previousEntry)
+
+            expect(await randomBeacon.isRelayRequestInProgress()).to.be.true
+          })
+        }
+      )
+
+      context("when no active groups exist after timeout is reported", () => {
+        let tx: ContractTransaction
+
+        beforeEach(async () => {
+          // `groupSize * relayEntrySubmissionEligibilityDelay +
+          // relayEntryHardTimeout`.
+          await mineBlocks(64 * 10 + 5760)
+
+          tx = await randomBeacon.reportRelayEntryTimeout()
+        })
+
+        it("should slash entire stakes of all group members", async () => {
+          await expect(tx)
+            .to.emit(staking, "Slashed")
+            .withArgs(to1e18(1000), signersAddresses)
+        })
+
+        it("should terminate the group", async () => {
+          // TODO: Implementation once `Groups` library is ready.
+        })
+
+        it("should emit RelayEntryTimedOut event", async () => {
+          await expect(tx)
+            .to.emit(randomBeacon, "RelayEntryTimedOut")
+            .withArgs(1)
+        })
+
+        it("should clean up current relay request data", async () => {
+          // TODO: Uncomment those assertions once termination is implemented.
+          // await expect(tx).to.not.emit(randomBeacon, "RelayEntryRequested")
+          // expect(await randomBeacon.isRelayRequestInProgress()).to.be.false
```

Comment:
Should we uncoment them?

Reply:
Not yet. I left a TODO here. We can uncomment once group termination is implemented.